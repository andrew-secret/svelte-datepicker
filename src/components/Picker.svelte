<script lang="ts">
  import type { IUtils } from "@date-io/core/IUtils";
  import { createEventDispatcher, tick } from "svelte";
  import DatePickerHeader from "./DatePickerHeader/DatePickerHeader.svelte";
  import DayPicker from "./DayPicker/DayPicker.svelte";
  import YearPicker from "./YearPicker/YearPicker.svelte";
  import { openView } from "./YearPicker/YearPickerStore";
  import "./global.css";

  const defaultMinDate = new Date("1900-01-01");
  const defaultMaxDate = new Date("2099-12-31");

  export let dateAdapter: IUtils<Date>;
  export let locale: string | object | undefined;
  export let value = dateAdapter.date() as Date;
  export let focusedDay = dateAdapter.date() as Date;
  export let minDate: Date = defaultMinDate;
  export let maxDate: Date = defaultMaxDate;

  const dispatch = createEventDispatcher();

  if (locale) {
    dateAdapter.locale = locale;
  }

  function onDaySelect(day: Date): void {
    dispatch("selectDay", day);
  }

  $: selectedDates = (Array.isArray(value) ? value : [value])
    .filter(Boolean)
    .map(
      (selectedDateItem) =>
        selectedDateItem && dateAdapter.startOfDay(selectedDateItem)
    );

  $: currentMonth = value ?? dateAdapter.date();
  $: currentMonthNumber = dateAdapter.getMonth(currentMonth);

  $: needMonthSwitch =
    Boolean(focusedDay) && !dateAdapter.isSameMonth(currentMonth, focusedDay);

  $: fullMonth = dateAdapter.getWeekArray(currentMonth);

  function changeFocusedDay(newFocusDay: Date): void {
    if (focusedDay !== null && dateAdapter.isSameDay(newFocusDay, focusedDay)) {
      return;
    }

    focusedDay = newFocusDay;

    needMonthSwitch =
      Boolean(focusedDay) && !dateAdapter.isSameMonth(currentMonth, focusedDay);

    currentMonth = needMonthSwitch
      ? dateAdapter.startOfMonth(focusedDay)
      : currentMonth;
  }

  const handleFocus = (focusedDay: Date): void => {
    changeFocusedDay(focusedDay);
  };

  // TODO: Clean up this function and make it more DRY
  async function handleKeyDown(event: KeyboardEvent) {
    const current = document.activeElement as HTMLElement;
    let items = [...document.querySelectorAll('.day[aria-hidden="false"]')];
    const currentIndex = current ? items.indexOf(current) : 0;

    /*
      If we we don't find a current on document or the
      year picker is open we don't handle event 
      delagation for the day picker.
    */
    if (!current || $openView === "year") {
      return;
    }

    let newIndex: number;

    switch (event.key) {
      case "ArrowUp":
        const previousWeek = dateAdapter.addDays(focusedDay, -7);
        const isWeekOutOfMinDateRange = dateAdapter.isSameMonth(
          previousWeek,
          dateAdapter.getPreviousMonth(minDate)
        );

        if (isWeekOutOfMinDateRange) {
          return;
        }

        // This could perform a month switch
        changeFocusedDay(dateAdapter.addDays(focusedDay, -7));
        if (needMonthSwitch) {
          newIndex = 0;
        } else {
          newIndex = (currentIndex + items.length - 7) % items.length;
        }

        handeKeyDownFocus(newIndex, current);
        event.preventDefault();
        break;
      case "ArrowDown":
        const nextWeek = dateAdapter.addDays(focusedDay, 7);
        const isWeekOutOfMaxDateRange = dateAdapter.isSameMonth(
          nextWeek,
          dateAdapter.getNextMonth(maxDate)
        );

        if (isWeekOutOfMaxDateRange) {
          return;
        }

        // This could perform a month switch
        changeFocusedDay(dateAdapter.addDays(focusedDay, 7));
        if (needMonthSwitch) {
          newIndex = 0;
        } else {
          newIndex = (currentIndex + items.length + 7) % items.length;
        }

        // This could perform a month switch
        handeKeyDownFocus(newIndex, current);
        event.preventDefault();
        break;
      case "ArrowLeft":
        if (dateAdapter.isSameDay(focusedDay, minDate)) {
          return;
        }

        // This could perform a month switch
        changeFocusedDay(dateAdapter.addDays(focusedDay, -1));
        if (needMonthSwitch) {
          newIndex = 0;
        } else {
          newIndex = (currentIndex + items.length - 1) % items.length;
        }

        // This could perform a month switch
        handeKeyDownFocus(newIndex, current);
        event.preventDefault();
        break;
      case "ArrowRight":
        if (dateAdapter.isSameDay(focusedDay, maxDate)) {
          return;
        }

        // This could perform a month switch
        changeFocusedDay(dateAdapter.addDays(focusedDay, 1));

        if (needMonthSwitch) {
          newIndex = 0;
        } else {
          newIndex = (currentIndex + items.length + 1) % items.length;
        }

        handeKeyDownFocus(newIndex, current);
        event.preventDefault();
        break;
      case "Tab":
        newIndex = currentIndex;

      case "Enter":
        newIndex = currentIndex;
      default:
        break;
    }
  }

  async function handeKeyDownFocus(index: number, activeElement: HTMLElement) {
    let newItems: Element[];

    activeElement.blur();

    if (needMonthSwitch) {
      // because of two DOM manipulation we have to use tick
      await tick();
      newItems = [...document.querySelectorAll('.day[aria-hidden="false"]')];
      (newItems[0] as any).focus();
    }

    if (!needMonthSwitch) {
      // because of two DOM manipulation we have to use tick
      await tick();
      newItems = [...document.querySelectorAll('.day[aria-hidden="false"]')];
      (newItems[index] as any).focus();
    }
  }

  function selectNextMonth() {
    currentMonth = dateAdapter.getNextMonth(currentMonth);
    focusedDay = currentMonth;

    onDaySelect(currentMonth);
  }

  function selectPreviousMonth() {
    currentMonth = dateAdapter.getPreviousMonth(currentMonth);
    focusedDay = currentMonth;

    onDaySelect(currentMonth);
  }

  async function selectYear(year: Date): Promise<void> {
    const yearNumber = dateAdapter.getYear(year);

    /*
      update focusedDay to prevent month switch
      back to previous selection with keyboard controls
    */
    focusedDay = value;
    value = dateAdapter.setYear(focusedDay, yearNumber);
    onDaySelect(value);

    openView.setOpenView("days");

    await tick();
    const selectedDay = document.querySelector(".selected");
    if (selectedDay) {
      (selectedDay as HTMLButtonElement).focus();
    }
  }
</script>

<svelte:window on:keydown={handleKeyDown} />

<div class="date-picker">
  <DatePickerHeader
    {dateAdapter}
    {currentMonth}
    {selectNextMonth}
    {selectPreviousMonth}
  />
  {#if $openView === "year"}
    <YearPicker {dateAdapter} {currentMonth} {selectYear} />
  {/if}

  {#if $openView === "days"}
    <DayPicker
      {dateAdapter}
      {fullMonth}
      {selectedDates}
      {currentMonthNumber}
      {focusedDay}
      {onDaySelect}
      {handleFocus}
    />
  {/if}
</div>

<style>
  .date-picker {
    width: 322px;
    margin: 0 auto;
    display: flex;
    max-height: 380px;
    overflow-x: hidden;
    flex-direction: column;
    background-color: var(--sdp-bg-color);
  }
</style>
