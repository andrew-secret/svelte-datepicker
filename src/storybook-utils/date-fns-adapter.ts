import LuxonAdapter from '@date-io/luxon';
import DateFnsAdapter from "@date-io/date-fns";

export const dateFnsUtils = new DateFnsAdapter();
export const luxonUtils = new LuxonAdapter();