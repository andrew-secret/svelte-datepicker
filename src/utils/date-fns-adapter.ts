import LuxonAdapter from '@date-io/luxon';
import DateFnsAdapter from "@date-io/date-fns";
import DayjsAdapter from '@date-io/dayjs'
import MomentUtils from '@date-io/moment'

// Should only be used in Storybook 
export const dateFnsUtils = new DateFnsAdapter();
export const luxonUtils = new LuxonAdapter();
export const dayJsUtils = new DayjsAdapter();
export const momentUtils = new MomentUtils();
