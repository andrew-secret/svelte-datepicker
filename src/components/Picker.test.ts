import '@testing-library/jest-dom/extend-expect';
import { render, fireEvent } from '@testing-library/svelte';
import { dateFnsUtils } from '../storybook-utils/date-fns-adapter';
import Picker from './Picker.svelte';

describe('Picker', () => {
    const value = new Date('2021-04-30');
    const props = {
        dateAdapter: dateFnsUtils,
        value,
        focusedDay: value,
    }

    it('should render w/ daypicker', () => {
        const { container } = render(Picker, props);
        expect(container).toMatchSnapshot();
    });
});