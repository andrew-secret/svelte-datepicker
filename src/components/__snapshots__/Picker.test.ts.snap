// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Picker should render w/ daypicker 1`] = `
<body>
  <div>
    <h1>
      Date: 
      Fri Apr 30 2021 02:00:00 GMT+0200 (Central European Summer Time)
      !
    </h1>
     
    <div
      class="date-picker svelte-1jmgq4v"
    >
      <div
        class="calendar-header svelte-jn3mp9"
      >
        <div
          class="year-switcher-wrapper"
        >
          <time
            aria-live="polite"
            class="year-switcher-label"
            data-testid="selected-month"
            datetime="Apr 1, 2021"
          >
            April
          </time>
           
          <time
            aria-live="polite"
            class="year-switcher-label"
            data-testid="selected-year"
            datetime="Apr 1, 2021"
          >
            2021
          </time>
           
          <button
            class="month-switcher svelte-jn3mp9"
            type="button"
          >
            <span
              class="month-switcher-label svelte-jn3mp9"
            >
              <svg
                aria-hidden="true"
                class="caret"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </span>
          </button>
        </div>
         
        <div
          class="month-switcher-wrapper"
        >
          <button
            aria-label="previous month"
            class="month-switcher previous-month svelte-jn3mp9"
          >
            <span
              class="month-switcher-label svelte-jn3mp9"
            >
              <svg
                aria-hidden="true"
                data-testid="ArrowLeftIcon"
                fill="currentColor"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
                />
              </svg>
            </span>
          </button>
           
          <button
            aria-label="next month"
            class="month-switcher next-month svelte-jn3mp9"
          >
            <span
              class="month-switcher-label svelte-jn3mp9"
            >
              <svg
                aria-hidden="true"
                data-testid="ArrowRightIcon"
                fill="currentColor"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
       
       
      <div
        class="weekdays svelte-bdh7zw"
      >
        <span
          class="weekday-label svelte-bdh7zw"
          title="Su"
        >
          Su
           
        </span>
        <span
          class="weekday-label svelte-bdh7zw"
          title="Mo"
        >
          Mo
           
        </span>
        <span
          class="weekday-label svelte-bdh7zw"
          title="Tu"
        >
          Tu
           
        </span>
        <span
          class="weekday-label svelte-bdh7zw"
          title="We"
        >
          We
           
        </span>
        <span
          class="weekday-label svelte-bdh7zw"
          title="Th"
        >
          Th
           
        </span>
        <span
          class="weekday-label svelte-bdh7zw"
          title="Fr"
        >
          Fr
           
        </span>
        <span
          class="weekday-label svelte-bdh7zw"
          title="Sa"
        >
          Sa
           
        </span>
      </div>
       
      <div
        role="grid"
      >
        <div
          class="week svelte-bdh7zw"
          role="row"
        >
          <button
            aria-hidden="true"
            aria-label="Sun Mar 28 2021 00:00:00 GMT+0100 (Central European Standard Time)"
            class="day svelte-bdh7zw outside-of-current-month"
            data-testid="hidden-day"
            role="cell"
          >
            28
          </button>
          <button
            aria-hidden="true"
            aria-label="Mon Mar 29 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw outside-of-current-month"
            data-testid="hidden-day"
            role="cell"
          >
            29
          </button>
          <button
            aria-hidden="true"
            aria-label="Tue Mar 30 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw outside-of-current-month"
            data-testid="hidden-day"
            role="cell"
          >
            30
          </button>
          <button
            aria-hidden="true"
            aria-label="Wed Mar 31 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw outside-of-current-month"
            data-testid="hidden-day"
            role="cell"
          >
            31
          </button>
          <button
            aria-hidden="false"
            aria-label="Thu Apr 01 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            1
          </button>
          <button
            aria-hidden="false"
            aria-label="Fri Apr 02 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            2
          </button>
          <button
            aria-hidden="false"
            aria-label="Sat Apr 03 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            3
          </button>
           
        </div>
        <div
          class="week svelte-bdh7zw"
          role="row"
        >
          <button
            aria-hidden="false"
            aria-label="Sun Apr 04 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            4
          </button>
          <button
            aria-hidden="false"
            aria-label="Mon Apr 05 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            5
          </button>
          <button
            aria-hidden="false"
            aria-label="Tue Apr 06 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            6
          </button>
          <button
            aria-hidden="false"
            aria-label="Wed Apr 07 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            7
          </button>
          <button
            aria-hidden="false"
            aria-label="Thu Apr 08 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            8
          </button>
          <button
            aria-hidden="false"
            aria-label="Fri Apr 09 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            9
          </button>
          <button
            aria-hidden="false"
            aria-label="Sat Apr 10 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            10
          </button>
           
        </div>
        <div
          class="week svelte-bdh7zw"
          role="row"
        >
          <button
            aria-hidden="false"
            aria-label="Sun Apr 11 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            11
          </button>
          <button
            aria-hidden="false"
            aria-label="Mon Apr 12 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            12
          </button>
          <button
            aria-hidden="false"
            aria-label="Tue Apr 13 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            13
          </button>
          <button
            aria-hidden="false"
            aria-label="Wed Apr 14 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            14
          </button>
          <button
            aria-hidden="false"
            aria-label="Thu Apr 15 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            15
          </button>
          <button
            aria-hidden="false"
            aria-label="Fri Apr 16 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            16
          </button>
          <button
            aria-hidden="false"
            aria-label="Sat Apr 17 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            17
          </button>
           
        </div>
        <div
          class="week svelte-bdh7zw"
          role="row"
        >
          <button
            aria-hidden="false"
            aria-label="Sun Apr 18 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            18
          </button>
          <button
            aria-hidden="false"
            aria-label="Mon Apr 19 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            19
          </button>
          <button
            aria-hidden="false"
            aria-label="Tue Apr 20 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            20
          </button>
          <button
            aria-hidden="false"
            aria-label="Wed Apr 21 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            21
          </button>
          <button
            aria-hidden="false"
            aria-label="Thu Apr 22 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            22
          </button>
          <button
            aria-hidden="false"
            aria-label="Fri Apr 23 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            23
          </button>
          <button
            aria-hidden="false"
            aria-label="Sat Apr 24 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            24
          </button>
           
        </div>
        <div
          class="week svelte-bdh7zw"
          role="row"
        >
          <button
            aria-hidden="false"
            aria-label="Sun Apr 25 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            25
          </button>
          <button
            aria-hidden="false"
            aria-label="Mon Apr 26 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            26
          </button>
          <button
            aria-hidden="false"
            aria-label="Tue Apr 27 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            27
          </button>
          <button
            aria-hidden="false"
            aria-label="Wed Apr 28 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            28
          </button>
          <button
            aria-hidden="false"
            aria-label="Thu Apr 29 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw"
            data-testid="visible-day"
            role="cell"
          >
            29
          </button>
          <button
            aria-hidden="false"
            aria-label="Fri Apr 30 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            autofocus=""
            class="day svelte-bdh7zw today selected"
            data-testid="visible-day"
            role="cell"
          >
            30
          </button>
          <button
            aria-hidden="true"
            aria-label="Sat May 01 2021 00:00:00 GMT+0200 (Central European Summer Time)"
            class="day svelte-bdh7zw outside-of-current-month"
            data-testid="hidden-day"
            role="cell"
          >
            1
          </button>
           
        </div>
      </div>
    </div>
  </div>
</body>
`;
